src/library-authoring/components/LibraryComponents.tsx

```diff
@@ -0,0 +1,141 @@
+import React, { useEffect, useMemo } from 'react';
+
+import { CardGrid } from '@openedx/paragon';
+import { NoComponents, NoSearchResults } from '../EmptyStates';
+import { useLibraryBlockTypes, useLibraryComponentCount, useLibraryComponents } from '../data/apiHook';
+import { ComponentCard, ComponentCardLoading } from './ComponentCard';
+
+type LibraryComponentsProps = {
+  libraryId: string,
+  filter: {
+    searchKeywords: string,
+  },
+  variant: string,
+};
+
+/**
+ * Library Components to show components grid
+ *
+ * Use style to:
+ *   - 'full': Show all components with Infinite scroll pagination.
+ *   - 'preview': Show first 4 components without pagination.
+ */
+const LibraryComponents = ({
+  libraryId,
+  filter: { searchKeywords },
+  variant,
+}: LibraryComponentsProps) => {
+  const { componentCount } = useLibraryComponentCount(libraryId, searchKeywords);
+  const {
+    hits,
+    isFetching,
+    isFetchingNextPage,
+    hasNextPage,
+    fetchNextPage,
+  } = useLibraryComponents(libraryId, searchKeywords);
+
+  const { componentList, tagCounts } = useMemo(() => {
+    const result = variant === 'preview' ? hits.slice(0, 4) : hits;
+    const tagsCountsResult = {};
+    result.forEach((component) => {
+      if (!component.tags) {
+        tagsCountsResult[component.id] = 0;
+      } else {
+        tagsCountsResult[component.id] = (component.tags.level0?.length || 0)
+              + (component.tags.level1?.length || 0)
+              + (component.tags.level2?.length || 0)
+              + (component.tags.level3?.length || 0);
+      }
+    });
+    return {
+      componentList: result,
+      tagCounts: tagsCountsResult,
+    };
+  }, [hits]);
```

Comment:
`componentList` is so simple that it doesn't need to be memoized.

Also, what you have here is fine but I think it would be nicer if you just compute the tagsCount inside `<ComponentCard>`. Instead of passing so many parameters to `<ComponentCard>` and processing them all in this function and memoizing some results, just pass a single `ContentHit` property to `ComponentCard`.

```suggestion
  const componentList = variant === 'preview' ? hits.slice(0, 4) : hits;

  const tagCounts = useMemo(() => {
    const tagsCountsResult = {};
    hits.forEach((component) => {
      if (!component.tags) {
        tagsCountsResult[component.id] = 0;
      } else {
        tagsCountsResult[component.id] = (component.tags.level0?.length || 0)
              + (component.tags.level1?.length || 0)
              + (component.tags.level2?.length || 0)
              + (component.tags.level3?.length || 0);
      }
    });
    return tagsCountsResult;
  }, [hits]);
```

Reply:
> just pass a single `ContentHit` property to `ComponentCard` 

I think that doing this would reduce the ability of `ComponentCard` to be reused, but for now I have not seen cases where it is used with parameters other than `ContentHit`, so I see it well too :+1:

Reply:
Updated https://github.com/openedx/frontend-app-course-authoring/pull/1109/commits/3d095351b84dcdb1318cdbe69775ae6edda81846